### ISSUE TITLE
Undefined names: missing imports?

### ISSUE THREAD

**Issue by cclauss** [d2l-ai/d2l-en#686]
[flake8](http://flake8.pycqa.org) testing of https://github.com/d2l-ai/d2l-en on Python 3.8.0

$ __flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics__
```
./d2l/d2l.py:361:5: F821 undefined name 'DATA_URL'
    DATA_URL+'kaggle_house_pred_train.csv',
    ^
./d2l/d2l.py:515:13: F821 undefined name 'read_time_machine'
    lines = read_time_machine()
            ^
./d2l/d2l.py:761:27: F821 undefined name 'read_data_nmt'
    text = preprocess_nmt(read_data_nmt())
                          ^
./d2l/d2l.py:1481:17: F821 undefined name 'read_ptb'
    sentences = read_ptb()
                ^
./d2l/d2l.py:1590:34: F821 undefined name 'read_data_ml100k'
    data, num_users, num_items = read_data_ml100k()
                                 ^
5     F821 undefined name 'DATA_URL'
5
```
https://flake8.pycqa.org/en/latest/user/error-codes.html

On the flake8 test selection, this PR does _not_ focus on "_style violations_" (the majority of flake8 error codes that [__psf/black__](https://github.com/psf/black) can autocorrect).  Instead these tests are focus on runtime safety and correctness:
* E9 tests are about Python syntax errors usually raised because flake8 can not build an Abstract Syntax Tree (AST).  Often these issues are a sign of unused code or code that has not been ported to Python 3.  These would be compile-time errors in a compiled language but in a dynamic language like Python they result in the script halting/crashing on the user.
* F63 tests are usually about the confusion between identity and equality in Python.  Use ==/!= to compare str, bytes, and int literals is the classic case.  These are areas where __a == b__ is True but __a is b__ is False (or vice versa).  Python >= 3.8 will raise SyntaxWarnings on these instances.
* F7 tests logic errors and syntax errors in type hints
* F82 tests are almost always _undefined names_ which are usually a sign of a typo, missing imports, or code that has not been ported to Python 3.  These also would be compile-time errors in a compiled language but in Python a __NameError__ is raised which will halt/crash the script on the user.

**Comment by astonzhang** [comment]
See https://github.com/d2l-ai/d2l-en/issues/685

**Comment by cclauss** [comment]
One down, four to go.

**Comment by Binh0103** [comment]
.

**Comment by astonzhang** [comment]
We have fixed the missing imports. For the remaining flake8 style issues, we'll fix in the end

**Comment by cclauss** [comment]
$ __flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics__
```
./d2l/d2l.py:1654:24: F821 undefined name 'vocab'
    premise = np.array(vocab[premise],
                       ^
./d2l/d2l.py:1656:27: F821 undefined name 'vocab'
    hypothesis = np.array(vocab[hypothesis],
                          ^
2     F821 undefined name 'vocab'
2
```

**Comment by cclauss** [comment]
Fixed in #953
